File: java/demo/impl/IntegerSequenceImpl.java
package demo.impl;

import java.util.Arrays;
import java.util.Date;

import demo.Response;

/**
 * The generic implementation of the integer sequence service.
 */
public abstract class IntegerSequenceImpl extends demo.IntegerSequencePOA {

	private final String name;
	
	private final String description;
	
	private final int maxIndex;
	
	protected IntegerSequenceImpl(String name, String description, int maxIndex) {
		this.name = name;
		this.description = description;
		this.maxIndex = maxIndex;
	}
	
	@Override
	public String name() {
		return name;
	}

	@Override
	public String description() {
		return description;
	}

	@Override
	public int maxIndex() {
		return maxIndex;
	}
	
	private Response get(int index) {
		Response response = new Response();
		try {
			if (index < 0) {
				throw new ServiceException("Index cannot be negative");
			} else if (index > this.maxIndex()) {
				throw new ServiceException("Index is too big");
			}
			
			response = this.compute(index);
		} catch (Exception e) {
			response.message(e.getMessage());
		}
		
		return response;
	}

	@Override
	public Response number(int index) {
		System.out.format("Requested number #%d from sequence '%s'\n", index, this.name());
		long tStart = new Date().getTime();
		Response response = this.get(index);
		System.out.format("Responded in %d ms\n", new Date().getTime() - tStart);
		return response;
	}

	@Override
	public Response[] numbers(int[] indices) {
		System.out.format("Requested number(s) #%s from sequence '%s'\n", 
				Arrays.toString(indices), this.name());

		long tStart = new Date().getTime();
		Response[] responses = new Response[indices.length];
		for (int i = 0; i < indices.length; i++) {
			responses[i] = this.get(indices[i]);
		}
		System.out.format("Responded in %d ms\n", new Date().getTime() - tStart);
		return responses;
	}
	
	protected abstract Response compute(int index) throws Exception;
	
	public abstract Name nameForCORBA();
}

File: cs/src/IntegerSequenceImpl.cs
using System;
using Ch.Elca.Iiop.Idl;

namespace demo.impl {

	/// <summary>
	/// The generic implementation of the integer sequence service.
	/// </summary>
	[SupportedInterface(typeof(demo.IntegerSequence))]
	public abstract class IntegerSequenceImpl : MarshalByRefObject, demo.IntegerSequence {

		protected IntegerSequenceImpl() {
		}

		public override object InitializeLifetimeService() {
			// Live forever
			return null;
		}

		public string name {
			get;
			protected set;
		}

		public int maxIndex {
			get;
			protected set;
		}

		public string description {
			get;
			protected set;
		}

		private Response GetNumber(int index) {
			try {
				if (index < 0) throw new IndexOutOfRangeException("Index cannot be negative");
				if (index > this.maxIndex) throw new IndexOutOfRangeException("Index is too big");

				return this.Compute(index);
			} catch (Exception e) {
				Console.WriteLine(e.Message);
				Response res = new Response();
				res.Setmessage(e.Message);
				return res;
			}
		}

		protected abstract Response Compute(int index);

		protected void Log(string message, params object[] args) {
			Console.WriteLine("[{0}] {1}", this, string.Format(message, args));
		}

		public Response number(int index) {
			Console.WriteLine("Requested number #{1} from sequence '{0}'", this, index);
			DateTime tStart = DateTime.Now;
			Response val = this.GetNumber(index);
			Console.WriteLine("Responded in {0} ms", (DateTime.Now - tStart).TotalMilliseconds);
			return val;
		}

		public Response[] numbers(int[] indices) {
			Console.WriteLine("Requested number(s) #[{1}] from sequence '{0}'", this, string.Join(", ", indices));

			DateTime tStart = DateTime.Now;
			Response[] res = new Response[indices.Length];
			for (int i = 0; i < indices.Length; i++) {
				res[i] = this.GetNumber(indices[i]);
			}
			Console.WriteLine("Responded in {0} ms", (DateTime.Now - tStart).TotalMilliseconds);

			return res;
		}

		public abstract Name CorbaName {
			get;
		}

		public override string ToString() {
			return this.name;
		}
	}
}
